README.md

```diff
@@ -1,86 +1,108 @@
 
-Let's consider a platform that receives data from many IoT devices. IoT devices have many models and different models send data with different fields. Generally these are stored in blob/nosql storage, a separate operation scans this storage for changes in schema.
+## Story
 
-Usually the changes in schema are additive only that is new fields in a new piece of data are added to schema. All records not having these fields automatically have these as `null` values.
+YAP has developed a sound business model where they let other companies upload their photographs to their dataset. When user search for the photographs they show up with the companies branding.
 
-The problem here is that the schema is generated and stored separately and not as part of the application code that will interact with the data. This means that either the schema has to be copied by hand or is only accessible at runtime. This significantly reduces the kind of validations that can be performed at runtime.
+The dataset submitted by third party providers runs through to pre-processing and vettings stages before being accepted. The pre-processing metadata is written to a persistent key value store, the analysis stage generates tags for the image which is commited to a graph database and rejected images are stored in a file storage to be sent back to provider. YAP wants to ensure the **sequence of steps is maintained across multiple components** (to avoid legal headaches obviously 😝). However the number and complexity of the integrations make it very difficult to maintain the test cases.
+
+Furthermore some third party vendors want to integrate the YAP search with their own applications. They want to create an image query and then get notified when any new images matching the query are uploaded. However each vendor have various limitations on the amount and rate of traffic they can handle, so YAP builds in rate limits and circuit breakers patterns for the different vendors. In this complex architecture, YAP finds it **difficult to reason about performance bottlenecks and when and where they can occur in the application**, especially when adding new integrations.
+
+Furthermore now YAP has become a multi-tenant search engine while itself running on top a multi-tenant infrastructure provider. YAP is facing considerable difficulty understanding the cost of onboarding new tenants, **modelling the cost of building new features** across tenants. This is further complicated cloud providers have granular charges (upto per request or per GB consumed cost) and tenants have diverse traffic patterns.
+
+### Idea
+
+The three problems mentioned above, namely reasoning about performance, enforcing invariants and reasing about cost are inter-related and point to the lack of analysis tools that reason across cloud components. Most of these properties can only be determined by through logs, traces and metrics generated by running the system which is costly and error prone.
```

Comment:
I would argue that while they are related you are tackling 3 different and difficult problems together in one section rather than splitting it up.

Reply:
Collecting in separate issue.